### PR TITLE
Gtk3 Updates

### PR DESCRIPTION
**Pull request by Bob-IT**
These changes fix some issues noticed in the new Gtk3 version.

### CODE REVIEW

File: gnucash/gnome/gnc-plugin-page-sx-list.c [hunk 1]
```diff
@@ -606,7 +606,8 @@ gnc_plugin_page_sx_list_cmd_new(GtkAction *action, GncPluginPageSxList *page)
         schedule = g_list_append(schedule, r);
         gnc_sx_set_schedule(new_sx, schedule);
     }
-    gnc_ui_scheduled_xaction_editor_dialog_create(new_sx, new_sx_flag);
+    gnc_ui_scheduled_xaction_editor_dialog_create(gnc_ui_get_main_window (NULL),
```
Comment (gjanssens) [inline gnucash/gnome/gnc-plugin-page-sx-list.c]: Plugin pages are always part of a main window somehow. So you can get their main_window via
```
GTK_WINDOW (gnc_plugin_page_get_window (GNC_PLUGIN_PAGE (plugin_page)));
```

 There should be no need to call gnc_ui_get_main_window here.

File: gnucash/gnome/gnc-plugin-page-sx-list.c [hunk 2]
```diff
@@ -638,7 +639,8 @@ gnc_plugin_page_sx_list_cmd_new2 (GtkAction *action, GncPluginPageSxList *page)
 static void
 _edit_sx(gpointer data, gpointer user_data)
 {
-    gnc_ui_scheduled_xaction_editor_dialog_create((SchedXaction*)data, FALSE);
+    gnc_ui_scheduled_xaction_editor_dialog_create(gnc_ui_get_main_window (NULL),
+        (SchedXaction*)data, FALSE);
```
Comment (gjanssens) [inline gnucash/gnome/gnc-plugin-page-sx-list.c]: Same here.

File: gnucash/gnome/gnc-plugin-page-sx-list.c [hunk 3]
```diff
@@ -722,7 +724,7 @@ gppsl_row_activated_cb(GtkTreeView *tree_view,
     GncPluginPageSxListPrivate *priv = GNC_PLUGIN_PAGE_SX_LIST_GET_PRIVATE(page);
 
     SchedXaction *sx = gnc_tree_view_sx_list_get_sx_from_path(GNC_TREE_VIEW_SX_LIST(priv->tree_view), path);
-    gnc_ui_scheduled_xaction_editor_dialog_create(sx, FALSE);
+    gnc_ui_scheduled_xaction_editor_dialog_create(gnc_ui_get_main_window (NULL), sx, FALSE);
```
Comment (gjanssens) [inline gnucash/gnome/gnc-plugin-page-sx-list.c]: Same here. You have the page at your disposal, use its main_window directly.

File: gnucash/gnome/gnc-split-reg.c [hunk 4]
```diff
@@ -1363,7 +1363,8 @@ gsr_default_schedule_handler( GNCSplitReg *gsr, gpointer data )
 
 	if ( theSX )
 	{
-	    gnc_ui_scheduled_xaction_editor_dialog_create(theSX, FALSE);
+	    gnc_ui_scheduled_xaction_editor_dialog_create(gnc_ui_get_main_window (NULL),
```
Comment (gjanssens) [inline gnucash/gnome/gnc-split-reg.c]: And here you can either query the main_window from the split register (which I believe has its own function to retrieve it) or you can pass the gsr widget as parameter to gnc_ui_get_main_window. NULL should only be used if there's absolutely not widget that can be used to climb up the widget hierarchy to a top-level window.

File: gnucash/gnome/gnc-plugin-business.c [hunk 5]
```diff
@@ -910,6 +910,10 @@ gnc_plugin_business_update_menus (GncPluginPage *plugin_page)
     if (!plugin_page || !GNC_IS_PLUGIN_PAGE(plugin_page))
         return;
 
+    // Check that this is a main window and not embedded sx
+    if (!GNC_IS_MAIN_WINDOW(plugin_page->window))
+        return;
+
```
Comment (gjanssens) [inline gnucash/gnome/gnc-plugin-business.c]: Interesting, I was not aware of this...

File: gnucash/gnome/gnc-plugin-budget.c [hunk 6]
```diff
@@ -284,7 +284,7 @@ gnc_budget_gui_select_budget(QofBook *book)
     gboolean ok;
 
     dlg = GTK_DIALOG(gtk_dialog_new_with_buttons(
-                         _("Select a Budget"), NULL, GTK_DIALOG_MODAL,
+                         _("Select a Budget"), gnc_ui_get_main_window(NULL), GTK_DIALOG_MODAL,
```
Comment (gjanssens) [inline gnucash/gnome/gnc-plugin-budget.c]: The two commands calling `gnc_budget_gui_select_budget` have access to a main_window pointer. This pointer should be passed as additional parameter to `gnc_budget_gui_select_budget`

File: gnucash/gnome/gnc-plugin-page-budget.c [hunk 7]
```diff
@@ -769,9 +769,7 @@ gnc_plugin_page_budget_cmd_view_options (GtkAction *action,
 
         priv->dialog = GTK_WIDGET(gtk_builder_get_object (builder, "budget_options_container_dialog"));
 
-        gtk_window_set_transient_for(
-            GTK_WINDOW(priv->dialog),
-            GTK_WINDOW(gnc_plugin_page_get_window(GNC_PLUGIN_PAGE(page))));
+        gtk_window_set_transient_for (GTK_WINDOW (priv->dialog), gnc_ui_get_main_window(NULL));
```
Comment (gjanssens) [inline gnucash/gnome/gnc-plugin-page-budget.c]: The original was correct :)

File: gnucash/gnome/gnc-plugin-page-budget.c [hunk 8]
```diff
@@ -923,6 +921,8 @@ gnc_plugin_page_budget_cmd_estimate_budget(GtkAction *action,
 
     dialog = GTK_WIDGET(gtk_builder_get_object (builder, "budget_estimate_dialog"));
 
+    gtk_window_set_transient_for (GTK_WINDOW (dialog), gnc_ui_get_main_window(NULL));
+
```
Comment (gjanssens) [inline gnucash/gnome/gnc-plugin-page-budget.c]: Here as well, reuse the main_window already available to the plugin page.

File: gnucash/gnome-utils/dialog-preferences.c [hunk 9]
```diff
@@ -1236,6 +1236,9 @@ gnc_preferences_dialog_create(void)
     // Set the style context for this dialog so it can be easily manipulated with css
     gnc_widget_set_style_context (GTK_WIDGET(dialog), "GncPreferenceDialog");
 
+    /* parent */
+    gtk_window_set_transient_for (GTK_WINDOW(dialog), GTK_WINDOW(gnc_ui_get_main_window (NULL)));
```
Comment (gjanssens) [inline gnucash/gnome-utils/dialog-preferences.c]: See if you can get the calling functions to pass a parent widget like for the others below.

**Comment by Bob-IT** [comment]
I have changed most of the transient parent warnings to use the page window and added some missed ones for the SX. Hopefully OK.
Bob

**Comment by jralls** [comment]
Thanks. It looks like you dealt with @gjanssens's comments so I went ahead and merged it to get it into 2.7.3. I do notice there are still a few places where the alignment doesn't look right. What are you using for an editor?

**Pull request merged**